kernel/comps/virtio/src/queue.rs

```diff
@@ -32,151 +32,115 @@ pub enum QueueError {
 
 /// The mechanism for bulk data transport on virtio devices.
 ///
-/// Each device can have zero or more virtqueues.
+/// A device can have zero or several virtqueues.
+///
+/// The const parameter `SIZE` specifies the size of the queue,
+/// correlating to the number of descriptors, and the number of slots
+/// in both the available and used rings.
 #[derive(Debug)]
-pub struct VirtQueue {
-    /// Descriptor table
-    descs: Vec<SafePtr<Descriptor, DmaCoherent>>,
-    /// Available ring
-    avail: SafePtr<AvailRing, DmaCoherent>,
-    /// Used ring
-    used: SafePtr<UsedRing, DmaCoherent>,
-    /// point to notify address
+pub struct VirtQueue<const SIZE: usize> {
+    /// The index of the queue
+    idx: u16,
+    /// The descriptor table
+    desc_table: DescTable,
+    /// The available ring
+    aval_ring: AvailRing<SIZE>,
+    /// The used ring
+    used_ring: UsedRing<SIZE>,
+    /// The notify pointer
     notify: SafePtr<u32, IoMem>,
-
-    /// The index of queue
-    queue_idx: u32,
-    /// The size of the queue.
-    ///
-    /// This is both the number of descriptors, and the number of slots in the available and used
-    /// rings.
-    queue_size: u16,
-    /// The number of used queues.
-    num_used: u16,
-    /// The head desc index of the free list.
-    free_head: u16,
-    /// the index of the next avail ring index
-    avail_idx: u16,
-    /// last service used index
-    last_used_idx: u16,
 }
 
-impl VirtQueue {
-    /// Create a new VirtQueue.
-    pub(crate) fn new(
-        idx: u16,
-        size: u16,
-        transport: &mut dyn VirtioTransport,
-    ) -> Result<Self, QueueError> {
-        if !size.is_power_of_two() {
+impl<const SIZE: usize> VirtQueue<SIZE> {
+    /// Creates a new `VirtQueue`.
+    pub(crate) fn new(idx: u16, transport: &mut dyn VirtioTransport) -> Result<Self, QueueError> {
+        if !SIZE.is_power_of_two()
+            || SIZE > u16::MAX as usize
+            || SIZE > transport.max_queue_size(idx).unwrap() as usize
```

Comment:
The driver cannot unilaterally determine the queue size at compile time. This is because the actual queue size must negotiated.

> Queue Size. On reset, specifies the maximum queue size supported by the device. This can be modified by the driver to reduce memory requirements. A 0 means the queue is unavailable.

So what the driver code probably can do is to statically determine the MAX queue size. But the actual queue size must still be negotiated.